test/testpushnotifications.cpp

```diff
@@ -0,0 +1,221 @@
+{
+    return QSharedPointer<OCC::PushNotifications>(new OCC::PushNotifications(account));
+}
+
+class TestPushNotifications : public QObject
+{
+    Q_OBJECT
+
+private slots:
+    void testSetup_correctCredentials_authenticatedOnWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+
+        QEventLoop loop;
+        QTimer::singleShot(5000, [&]() {
+            loop.quit();
+            QFAIL("No messages arrived");
+        });
+
+        bool userSended = false;
+        MockWebSocketServer mockServer;
+        connect(&mockServer, &MockWebSocketServer::processTextMessage, [&](QWebSocket *socket, QString message) {
+            if (message == user) {
+                userSended = true;
+                return;
+            } else if (message == password) {
+                QCOMPARE(userSended, true);
+                loop.quit();
+                return;
+            }
+            QFAIL("Unexpected message");
+        });
+
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+        pushNotifications->setup();
+
+        loop.exec();
+    }
+
+    void testOnWebSocketTextMessageReceived_notifyFileMessage_emitFilesChanged()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+
+        connect(&mockServer, &MockWebSocketServer::processTextMessage, [&](QWebSocket *sender, QString message) {
+            if (message == password) {
+                sender->sendTextMessage("notify_file");
+                return;
+            }
+        });
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        QEventLoop loop;
+        QTimer::singleShot(5000, [&]() {
+            loop.quit();
+            QFAIL("No files changed event emitted");
+        });
+
+        connect(pushNotifications.data(), &OCC::PushNotifications::filesChanged, [&](OCC::Account * /*account*/) {
+            loop.quit();
+        });
+
+        pushNotifications->setup();
+
+        loop.exec();
+    }
+
+    void testOnWebSocketTextMessageReceived_invalidCredentialsMessage_reconnectWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+
+        QEventLoop loop;
+        QTimer::singleShot(5000, [&]() {
+            loop.quit();
+            QFAIL("No retry");
+        });
+
+        bool userSended = false;
+        MockWebSocketServer mockServer;
+        quint16 passwordSendedCount = 0;
+        connect(&mockServer, &MockWebSocketServer::processTextMessage, [&](QWebSocket *socket, QString message) {
```

Comment:
By looking at that test and some others, I think your fake server misses some kind of "waitForTextMessage()" or similar which would receive a lambda returning a boolean. It would then get out of the event loop when the lambda returns true or when a timeout has been reached.

For the timeouts I'd advise against using QTimer::singleShot, especially the way you did it because those timers will fire off eventually (here if your whole test suite execution time would go beyond the five seconds I'd expect nice crashes to happen). So create and delete QTimer instances instead so that you can control when they run and when they die.